File: android/app/src/main/java/com/theoplayer/contentprotectionintegration/integration/vudrm/VudrmWidevineContentProtectionIntegration.java
package com.theoplayer.contentprotectionintegration.integration.vudrm;

import com.theoplayer.android.api.contentprotection.CertificateRequestCallback;
import com.theoplayer.android.api.contentprotection.ContentProtectionIntegration;
import com.theoplayer.android.api.contentprotection.LicenseRequestCallback;
import com.theoplayer.android.api.contentprotection.Request;
import com.theoplayer.android.api.source.drm.DRMConfiguration;
import com.theoplayer.contentprotectionintegration.TypeUtils;

import org.json.JSONException;
import org.json.JSONObject;

public class VudrmWidevineContentProtectionIntegration extends ContentProtectionIntegration {

    static final String DEFAULT_LICENSE_URL = "https://widevine-proxy.drm.technology/proxy";

    private final DRMConfiguration contentProtectionConfiguration;

    public VudrmWidevineContentProtectionIntegration(DRMConfiguration configuration) {
        this.contentProtectionConfiguration = configuration;
    }

    private Request wrapRequest(Request request) {
        if (request.getBody() == null) {
            throw new NullPointerException("The request body can not be null");
        }
        final Object token = this.contentProtectionConfiguration.getIntegrationParameters().get("token");
        if (token == null) {
            throw new NullPointerException("The Widevine vuDRM token can not be null");
        }

        String licenseUrl = DEFAULT_LICENSE_URL;
        if (contentProtectionConfiguration.getWidevine() != null) {
            licenseUrl = contentProtectionConfiguration.getWidevine().getLicenseAcquisitionURL();
        }
        request.setUrl(licenseUrl);
        request.getHeaders().put("Content-Type", "text/plain");

        final Object kid = contentProtectionConfiguration.getIntegrationParameters().get("keyId");
        JSONObject jsonBody = new JSONObject();
        try {
            jsonBody.put("token", token);
            jsonBody.put("drm_info", TypeUtils.fromByteArrayToUint8JsonArray(request.getBody()));
            jsonBody.put("kid", kid);
        } catch (JSONException e) {
            e.printStackTrace();
        }
        request.setBody(TypeUtils.fromJsonToByteArray(jsonBody));
        return request;
    }

    public void onCertificateRequest(Request request, final CertificateRequestCallback callback) {
        try {
            callback.request(wrapRequest(request));
        } catch (Throwable error) {
            callback.error(error);
        }
    }

    public void onLicenseRequest(final Request request, LicenseRequestCallback callback) {
        try {
            callback.request(wrapRequest(request));
        } catch (Throwable error) {
            callback.error(error);
        }
    }
}

File: android/app/src/main/java/com/theoplayer/contentprotectionintegration/integration/vudrm/VudrmWidevineContentProtectionIntegration.kt
package com.theoplayer.contentprotectionintegration.integration.vudrm

import com.theoplayer.contentprotectionintegration.TypeUtils.fromByteArrayToUint8JsonArray
import com.theoplayer.contentprotectionintegration.TypeUtils.fromJsonToByteArray
import com.theoplayer.android.api.source.drm.DRMConfiguration
import com.theoplayer.android.api.contentprotection.ContentProtectionIntegration
import org.json.JSONObject
import org.json.JSONException
import com.theoplayer.android.api.contentprotection.CertificateRequestCallback
import com.theoplayer.android.api.contentprotection.LicenseRequestCallback
import com.theoplayer.android.api.contentprotection.Request
import java.lang.NullPointerException

class VudrmWidevineContentProtectionIntegration(private val contentProtectionConfiguration: DRMConfiguration) :
    ContentProtectionIntegration() {

    companion object {
        const val DEFAULT_LICENSE_URL = "https://widevine-proxy.drm.technology/proxy"
    }

    private fun wrapRequest(request: Request): Request {
        if (request.body == null) {
            throw NullPointerException("The request body can not be null")
        }
        val token = contentProtectionConfiguration.integrationParameters["token"]
            ?: throw NullPointerException("The Widevine vuDRM token can not be null")
        var licenseUrl = DEFAULT_LICENSE_URL
        if (contentProtectionConfiguration.widevine != null) {
            licenseUrl = contentProtectionConfiguration.widevine!!.licenseAcquisitionURL
        }
        request.url = licenseUrl
        request.headers["Content-Type"] = "text/plain"
        val kid = contentProtectionConfiguration.integrationParameters["keyId"]
        val jsonBody = JSONObject()
        try {
            jsonBody.put("token", token)
            jsonBody.put("drm_info", fromByteArrayToUint8JsonArray(request.body!!))
            jsonBody.put("kid", kid)
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        request.body = fromJsonToByteArray(jsonBody)
        return request
    }

    override fun onCertificateRequest(request: Request, callback: CertificateRequestCallback) {
        try {
            callback.request(wrapRequest(request))
        } catch (error: Throwable) {
            callback.error(error)
        }
    }

    override fun onLicenseRequest(request: Request, callback: LicenseRequestCallback) {
        try {
            callback.request(wrapRequest(request))
        } catch (error: Throwable) {
            callback.error(error)
        }
    }
}